include/directx/d3dx12.h

```diff
@@ -4037,6 +4037,665 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+    // Must be updated whenever a new root signature version is added to the d3d12.h header
+    HRESULT QueryHighestRootSignatureVersion()
+    {
+        HRESULT result;
+
+        D3D_ROOT_SIGNATURE_VERSION allRootSignatureVersions[] = {
+            D3D_ROOT_SIGNATURE_VERSION_1_1,
+            D3D_ROOT_SIGNATURE_VERSION_1_0,
+            D3D_ROOT_SIGNATURE_VERSION_1,
+        };
+
+        UINT numRootSignatureVersions = sizeof(allRootSignatureVersions) / sizeof(D3D_ROOT_SIGNATURE_VERSION);
+
+
+        // ROOT_SIGNATURE_VERSION_1_X
+        for (UINT i = 0; i < numRootSignatureVersions; i++) {
+            m_dRootSignature.HighestVersion = allRootSignatureVersions[i];
+            result = m_pDevice->CheckFeatureSupport(D3D12_FEATURE_ROOT_SIGNATURE, &m_dRootSignature, sizeof(D3D12_FEATURE_DATA_ROOT_SIGNATURE));
+            if (result != E_INVALIDARG) {
+                if (FAILED(result)) {
+                    m_dRootSignature.HighestVersion = (D3D_ROOT_SIGNATURE_VERSION)0;
+                }
+                // If succeeded, the highest version is already written into the member struct
+                return result;
+            }
+        }
+
+        // No version left. Set to invalid value and continue.
+        m_dRootSignature.HighestVersion = (D3D_ROOT_SIGNATURE_VERSION)0;
+        return S_OK;
+    }
+
+    // Helper funcion to decide the highest feature level
+    HRESULT QueryHighestFeatureLevel()
+    {
+        HRESULT result;
+
+        // Check against a list of all feature levels present in d3dcommon.h
+        // Needs to be updated for future feature levels
+        D3D_FEATURE_LEVEL allLevels[] = {
+            D3D_FEATURE_LEVEL_12_2,
+            D3D_FEATURE_LEVEL_12_1,
+            D3D_FEATURE_LEVEL_12_0,
+            D3D_FEATURE_LEVEL_11_1,
+            D3D_FEATURE_LEVEL_11_0,
+            D3D_FEATURE_LEVEL_10_1,
+            D3D_FEATURE_LEVEL_10_0,
+            D3D_FEATURE_LEVEL_9_3,
+            D3D_FEATURE_LEVEL_9_2,
+            D3D_FEATURE_LEVEL_9_1,
+            D3D_FEATURE_LEVEL_1_0_CORE
+        };
+
+        D3D12_FEATURE_DATA_FEATURE_LEVELS dFeatureLevel;
+        dFeatureLevel.NumFeatureLevels = sizeof(allLevels) / sizeof(D3D_FEATURE_LEVEL);
+        dFeatureLevel.pFeatureLevelsRequested = allLevels;
+
+        result = m_pDevice->CheckFeatureSupport(D3D12_FEATURE_FEATURE_LEVELS, &dFeatureLevel, sizeof(D3D12_FEATURE_DATA_FEATURE_LEVELS));
+        if (SUCCEEDED(result)) {
+            m_eMaxFeatureLevel = dFeatureLevel.MaxSupportedFeatureLevel;
+        } else {
+            m_eMaxFeatureLevel = (D3D_FEATURE_LEVEL)0;
+
+            if (result == DXGI_ERROR_UNSUPPORTED) { // Indicates that none are supported
+                result = S_OK;
+            }
+        }
+        return result;
+    }
+
+    // Helper function to initialize local protected resource session types structs
+    HRESULT QueryProtectedResourceSessionTypes(UINT NodeIndex, UINT Count)

```

Comment:
This is perhaps the only special check routine for part 2. 

1. It is one of the features that requires one feature data struct for each graphics node.
2. One of its in parameters depends on the results of another feature check, ProtectedResourceSessionTypeCount. Thus this check must be completed after getting the count in Init().
2. CheckFeatureSupport will take in a pointer (pTypes) and a size (Count) and store the output as a (C-style) array into the given pointer. The size is decidable from the ProtectedResourceSessionTypeCount before the API call, but it's probably not known at compile time. 

To avoid manually managing dynamic memory allocation, I've decided to extend the original D3D12_FEATURE_SUPPORT_DATA_PROTECTED_RESOURCE_SESSION_TYPES and add a vector to it. Upon initialization I will resize the vector to fit the type count and pass the underlying array's pointer (through vector::data()) to pTypes. When CheckFeatureSupport writes the output data to pTypes, it's essentially writing data to the vector. 

Finally, in the new checking API I simply returns the vector.